activesupport/lib/active_support/event_reporter.rb

```diff
@@ -0,0 +1,488 @@
+# typed: true
+# frozen_string_literal: true
+
+require_relative "event_reporter/encoders"
+
+module ActiveSupport
+  class TagStack # :nodoc:
+    EMPTY_TAGS = {}.freeze
+    FIBER_KEY = :event_reporter_tags
+
+    class << self
+      def tags
+        Fiber[FIBER_KEY] || EMPTY_TAGS
+      end
+
+      def with_tags(*args, **kwargs)
+        existing_tags = tags
+        tags = existing_tags&.dup || {}
+        tags.merge!(resolve_tags(args, kwargs))
+        new_tags = tags.freeze
+
+        begin
+          Fiber[FIBER_KEY] = new_tags
+          yield
+        ensure
+          Fiber[FIBER_KEY] = existing_tags
+        end
+      end
+
+      private
+        def resolve_tags(args, kwargs)
+          tags = args.each_with_object({}) do |arg, tags|
+            case arg
+            when String, Symbol
+              tags[arg.to_sym] = true
+            when Hash
+              arg.each { |key, value| tags[key.to_sym] = value }
+            else
+              tags[arg.class.name.to_sym] = arg
+            end
+          end
+          kwargs.each { |key, value| tags[key.to_sym] = value }
+          tags
+        end
+    end
+  end
+
+  class EventContext # :nodoc:
+    EMPTY_CONTEXT = {}.freeze
+    FIBER_KEY = :event_reporter_context
+
+    class << self
+      def context
+        Fiber[FIBER_KEY] || EMPTY_CONTEXT
+      end
+
+      def set_context(context_hash)
+        new_context = self.context.dup
+        context_hash.each { |key, value| new_context[key.to_sym] = value }
+
+        Fiber[FIBER_KEY] = new_context.freeze
+      end
+
+      def clear
+        Fiber[FIBER_KEY] = EMPTY_CONTEXT
+      end
+    end
+  end
+
+  # = Active Support \Event Reporter
+  #
+  # +ActiveSupport::EventReporter+ provides an interface for reporting structured events to subscribers.
+  #
+  # To report an event, you can use the +notify+ method:
+  #
+  #   Rails.event.notify("user_created", { id: 123 })
+  #   # Emits event:
+  #   #  {
+  #   #    name: "user_created",
+  #   #    payload: { id: 123 },
```

Comment:
Maybe out of scope for now, I just had this thought, but should we care about filtered parameters in this?

As we are cautious about those when logging, this seems another source of trouble.

Reply:
Yes, definitely 🙇‍♀️ I'll take a look at this.

Reply:
I took a stab at two ideas I had for this in Shopify/rails#37 and Shopify/rails#38.

Each has trade-offs, which depends on what we want to filter (e.g. just the payload, or the whole event).

FYI, I also added the filters to `ActiveSupport.filter_parameters` rather than storing it on the `EventReporter` instance, because it maybe useful outside this context (e.g. ErrorReporter).

Reply:
Super appreciated, I'll take a look! Probably does make sense to tackle independently, this PR is getting a bit large.